initialization/vrf_search.go

```diff
@@ -0,0 +1,183 @@
+	)
+
+	allFiles, err := os.ReadDir(initOpts.DataDir)
+	if err != nil {
+		return 0, nil, fmt.Errorf("couldn't open the data directory: %w", err)
+	}
+
+	woReference, err := oracle.New(
+		oracle.WithProviderID(CPUProviderID()),
+		oracle.WithCommitment(metadata.CommitmentAtxId),
+		oracle.WithVRFDifficulty(difficulty),
+		oracle.WithScryptParams(initOpts.Scrypt),
+		oracle.WithLogger(logger),
+	)
+	if err != nil {
+		return 0, nil, err
+	}
+	defer woReference.Close()
+
+	for _, file := range allFiles {
+		info, err := file.Info()
+		if err != nil {
+			logger.Error("failed to get file info", zap.Error(err))
+			continue
+		}
+
+		if !shared.IsInitFile(info) {
+			continue
+		}
+
+		fileIndex, err := shared.ParseFileIndex(file.Name())
+		if err != nil {
+			logger.Panic("failed to parse file index", zap.Error(err))
+		}
+		firstLabelIndex := uint64(fileIndex) * metadata.MaxFileSize / postrs.LabelLength
+		logger.Info("looking for VRF nonce in file", zap.String("file", file.Name()))
+
+		file, err := os.Open(filepath.Join(initOpts.DataDir, file.Name()))
+		if err != nil {
+			logger.Error("failed to open file", zap.Error(err))
+			continue
+		}
+		defer file.Close()
+
+		idx, l, err := searchForNonce(ctx, bufio.NewReader(file), difficulty, woReference)
+		if err != nil {
+			return 0, nil, fmt.Errorf("failed to search for nonce: %w", err)
+		}
+		if l != nil {
+			nonceValue = l
+			nonce = firstLabelIndex + idx
+			logger.Info("found nonce", zap.Uint64("index", nonce), zap.String("label", hex.EncodeToString(nonceValue)))
+			difficulty = nonceValue // override difficulty to the new lowest label
+
+		}
+	}
+	if nonceValue != nil {
+		logger.Info("updating postdata_metadata.json with found nonce", zap.Uint64("nonce", nonce), zap.String("NonceValue", hex.EncodeToString(nonceValue)))
+		metadata.Nonce = &nonce
+		metadata.NonceValue = shared.NonceValue(nonceValue)
+		SaveMetadata(initOpts.DataDir, metadata)
+		return nonce, nonceValue, nil
+	}
+	return 0, nil, ErrNonceNotFound
+}
+
+// searchForNonce searches for a nonce in the given reader.
+func searchForNonce(ctx context.Context, r io.Reader, difficulty []byte, oracle *oracle.WorkOracle) (nonce uint64, nonceValue []byte, err error) {
+	labelBuf := make([]byte, postrs.LabelLength)
+	labelIndex := uint64(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return nonce, labelBuf, ctx.Err()
+		default:
+			// continue looking for a nonce
+		}
+
+		_, err := io.ReadFull(r, labelBuf)
+		switch {
+		case err == io.EOF:
+			return nonce, nonceValue, nil
+		case err == io.ErrUnexpectedEOF:
+			return 0, nil, errors.New("unexpected end of file - file is truncated. please reinit it")
+		}
+
+		ok, err := checkLabel(labelIndex, labelBuf, difficulty, oracle)
+		if err != nil {
+			return 0, nil, fmt.Errorf("checking label: %w", err)
+		}
+		if ok {
+			nonce = labelIndex
+			nonceValue = append(nonceValue[:0], labelBuf...)
+			difficulty = nonceValue // override difficulty to the new lowest label
```

Comment:
Maybe also `copy` just to be safe?

Reply:
Do you mean to not modify `difficulty` by mistake by modifying nonceValue?

Reply:
yes. Just as a safety measure, in terms of performance it should make little to no difference.